libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -0,0 +1,343 @@
+// on success returns true and updates margin
+bool AP_OABendyRuler::calc_margin_from_circular_fence(const Location &start, const Location &end, float &margin)
+{
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr == nullptr) {
+        return false;
+    }
+    if ((fence_ptr->get_enabled_fences() & AC_FENCE_TYPE_CIRCLE) == 0) {
+        return false;
+    }
+
+    // calculate start and end point's distance from home
+    const Location ahrs_home = AP::ahrs().get_home();
+    const float start_dist = ahrs_home.get_distance_NE(start).length();
+    const float end_dist = ahrs_home.get_distance_NE(end).length();
+
+    // get circular fence radius
+    const float fence_radius = fence_ptr->get_radius();
+
+    // margin is fence radius minus the longer of start or end distance
+    margin = fence_radius - MAX(start_dist, end_dist);
+    return true;
+}
+
+// calculate minimum distance between a path and the polygon fence
+// on success returns true and updates margin
+bool AP_OABendyRuler::calc_margin_from_polygon_fence(const Location &start, const Location &end, float &margin)
+{
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr == nullptr) {
+        return false;
+    }
+    if (((fence_ptr->get_enabled_fences() & AC_FENCE_TYPE_POLYGON) == 0) || !fence_ptr->is_polygon_valid()) {
+        return false;
+    }
+
+    // get polygon boundary
+    // Note: first point in list is the return-point (which copter does not use)
+    uint16_t num_points;
+    const Vector2f* boundary = fence_ptr->get_polygon_points(num_points);
+    if (num_points < 4) {
+        // this should have already been checked by is_polygon_valid() but just in case
+        return false;
+    }
+
+    // convert start and end to offsets from EKF origin
+    Vector2f start_NE, end_NE;
+    if (!start.get_vector_xy_from_origin_NE(start_NE) || !end.get_vector_xy_from_origin_NE(end_NE)) {
+        return false;
+    }
+
+    // calculate min distance (in meters) from line to polygon
+    margin = Polygon_closest_distance_line(boundary, num_points, start_NE, end_NE) * 0.01f;
+    return true;
+}
+
+// calculate minimum distance between a path and proximity sensor obstacles
+// on success returns true and updates margin
+bool AP_OABendyRuler::calc_margin_from_proximity_sensors(const Location &start, const Location &end, float &margin)
+{
+    AP_Proximity* prx_ptr = AP_Proximity::get_singleton();
+    if (prx_ptr == nullptr) {
+        return false;
+    }
+
+    // retrieve obstacles from proximity library
+    uint16_t count = 0;
+    if (prx_ptr->copy_locations(_prx_locs, ARRAY_SIZE(_prx_locs), count)) {
+        _prx_loc_count = count;
+    }
+
+    // even on failure to retrieve latest locations (which can happen due to a semaphore conflict)
```

Comment:
How does this ever fail due to a semaphore conflict? It's a blocking take on the semaphore, so as far as I can tell this comment is wrong/misleading. The only time `copy_locations` fails is if we give it a null pointer, which 

A: we shouldn't be, passing by reference would be safer
B: if we did somehow then we shouldn't be continuing anyways

Reply:
ok, i've changed it to return false if it can't copy_locations.  also removed the unnecessary "count" local variable.